digital-asset/daml
# sandbox: PostgreSQL health checks.

Iterates on https://github.com/digital-asset/daml/issues/2733.

Sorry about the size of this one. On the upside, it's mostly test cases and small fixes to warnings I came across on the way.

This adds health checks to the Sandbox for the "index" and "write" services, and to the DAML-on-X Reference v2 server for "index", "read" and "write".

The "index" component will report as `NOT_SERVING` if more than 3 JDBC database operations fail to connect in a row.

On Sandbox, the "write" service will do the same.

On Reference v2, the "read" and "write" services will always report as healthy, because they run in-memory.

If no service name is specified, the health is an aggregate of all health checks; if any check reports that the service is unhealthy, we report NOT_SERVING.

This adds a mandatory `currentHealth(): HealthStatus` method to `IndexService`, `ReadService` and `WriteService`, among other traits. I'm not sure whether this should be defaulted to `Healthy`. The same applies to the mandatory `HealthChecks` parameter to `ApiServices.create`.

### Changelog

Reproduced here for ease-of-reading.

- [Ledger API Server] Add a health check endpoint conforming to the [GRPC Health Checking Protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md).
- [Ledger API Server] Add health checks for database connectivity.
- [Participant State API] Add a mandatory `currentHealth()` method to `IndexService`, `ReadService` and `WriteService`.

### Testing

You can verify this for yourself against the Sandbox by doing the following:

1. `docker run --rm -it -p 5432:5432 -e POSTGRES_USER=damlsandbox postgres`
2. `bazel run //ledger/sandbox:sandbox-binary -- --ledgerid=legdendary --port=6865 --eager-package-loading --sql-backend-jdbcurl='jdbc:postgresql://localhost:5432/damlsandbox?user=damlsandbox' $PWD/bazel-bin/ledger/test-common/*.dar`
3. `go get github.com/grpc-ecosystem/grpc-health-probe`
4. `${GOPATH:-~/go}/bin/grpc-health-probe -addr=localhost:6865`
5. Shut down PostgreSQL.
6. Hit the ledger API a few times.
7. Run the health probe again.
8. Start PostgreSQL back up.
9. Hit the API once more.
10. Run the health probe yet again.

You can pass a `-service` flag to `grpc-health-probe` to test individual services.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.
